workflow/controller/dag.go

```diff
@@ -434,81 +434,82 @@ func (woc *wfOperationCtx) executeDAGTask(ctx context.Context, dagCtx *dagContex
 		return
 	}
 
-	// Next, expand the DAG's withItems/withParams/withSequence (if any). If there was none, then
-	// expandedTasks will be a single element list of the same task
-	expandedTasks, err := expandTask(*newTask)
-	if err != nil {
-		woc.initializeNode(nodeName, wfv1.NodeTypeSkipped, dagTemplateScope, task, dagCtx.boundaryID, wfv1.NodeError, err.Error())
-		connectDependencies(nodeName)
-		return
-	}
+	taskNodeName := dagCtx.taskNodeName(newTask.Name)
+	node = dagCtx.getTaskNode(newTask.Name)
+	var expandedTasks []wfv1.DAGTask
+	if node == nil {
```

Comment:
I'm not comfortable with the refactor in this file. I don't think we should have so much (critical) logic controlled by nested if else statements. Is there a way to refactor it such that the `shouldExecute` check happens outside in its own block? Is it possible to have it as a normal block before the `expandedTasks, err := expandTask(*newTask)` line as it currently exists on the master branch?